sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala

```
@@ -0,0 +1,164 @@
+
+  override def requiredChildDistribution: Seq[Distribution] =
+    ClusteredOrderedDistribution(leftKeys) :: ClusteredOrderedDistribution(rightKeys) :: Nil
+
+  private val orders: Seq[SortOrder] = leftKeys.zipWithIndex.map {
+    case(expr, index) => SortOrder(BoundReference(index, expr.dataType, expr.nullable), Ascending)
+  }
+  private val ordering: RowOrdering = new RowOrdering(orders, left.output)
+
+  @transient protected lazy val leftKeyGenerator = newProjection(leftKeys, left.output)
+  @transient protected lazy val rightKeyGenerator = newProjection(rightKeys, right.output)
+
+  override def execute() = {
+    val leftResults = left.execute().map(_.copy())
+    val rightResults = right.execute().map(_.copy())
+
+    leftResults.zipPartitions(rightResults) { (leftIter, rightIter) =>
+      new Iterator[Row] {
+        // Mutable per row objects.
+        private[this] val joinRow = new JoinedRow5
+        private[this] var leftElement: Row = _
+        private[this] var rightElement: Row = _
+        private[this] var leftKey: Row = _
+        private[this] var rightKey: Row = _
+        private[this] var leftMatches: CompactBuffer[Row] = _
+        private[this] var rightMatches: CompactBuffer[Row] = _
+        private[this] var leftPosition: Int = -1
+        private[this] var rightPosition: Int = -1
+
+        override final def hasNext: Boolean = leftPosition != -1 || nextMatchingPair
+
+        override final def next(): Row = {
+          if (hasNext) {
+            val joinedRow = joinRow(leftMatches(leftPosition), rightMatches(rightPosition))
+            rightPosition += 1
+            if (rightPosition >= rightMatches.size) {
+              leftPosition += 1
+              rightPosition = 0
+              if (leftPosition >= leftMatches.size) {
+                leftPosition = -1
+              }
+            }
+            joinedRow
+          } else {
+            // according to Scala doc, this is undefined
+            null
```

I am still think it's better to throw an exception instead of returning a `null`, `null` is meaningful in some scenarios, which probably lead to troubles in debugging for some other modules.